hail/python/hail/experimental/function.py

```diff
@@ -2,21 +2,43 @@
 from hail.ir import Apply, Ref, Renderer, register_session_function
 from hail.expr.types import HailType
 from hail.expr.expressions import construct_expr, anytype, expr_any, unify_all
-from hail.typecheck import typecheck
+from hail.typecheck import typecheck, nullable
+import hail
 
+from decorator import decorator
 
 class Function(object):
-    def __init__(self, f, name):
+    def __init__(self, f, param_types, ret_type, name):
         self._f = f
         self._name = name
+        self._param_types = param_types
+        self._ret_type = ret_type
 
     def __call__(self, *args):
         return self._f(*args)
 
 
-@typecheck(f=anytype, param_types=HailType)
-def define_function(f, *param_types):
-    mname = Env.get_uid()
+@typecheck(param_types=hail.expr.types.hail_type)
```

Comment:
Why the inconsistency between this and `define_function` itself?

Reply:
HailType vs hail_type? The latter is a typechecker/coercer that will accept strings. I'll change the other one.